### ISSUE TITLE
Performance issue: Order of resizing/padding

### ISSUE THREAD

**Issue by felixplum** [eriklindernoren/PyTorch-YOLOv3#259]
Currently the image is first padded to its longest dimension, then resized. Especially for larger images the performance drop is substantial, as the biggest chunk of training time will be spent preprocessing images. In my case, I could reduce preprocessing time form 1.5s to 0.1s per size-32 batch.

The `__getItem__() `method in the `datasets.py` will just read the labels and store them, while `collate_fn() `calls the method below (add your own conversion routine, depending on how you labeled your images)

```python3
def convert_to_padded_square(images, targets=None, size=416):

    ## Do batch resizing and padding
    n_imgs = len(images)
    shapes_scaled = []
    # compute shapes of scaled images, with longer side = desired size
    for i in range(n_imgs):
        shape2d = images[i].shape[1:3] # height/width
        long_side = shape2d[0] if shape2d[0]>shape2d[1] else shape2d[1]
        ratio = size/long_side
        shapes_scaled += [(int((shape2d[0]+1.)*ratio),int((shape2d[1]+1.)*ratio))]
    # resize to shape
    
    imgs_resized = [resize(img, shape) for (img, shape) in zip(images, shapes_scaled)]
    # add paddings & keep track to adjust labels
    shapes_before_pad = torch.zeros(n_imgs, 2) # TODO: reuse values from above
    paddings = torch.zeros(n_imgs, 2) 
    for i in range(n_imgs):
        _, h, w = imgs_resized[i].shape
        shapes_before_pad[i, 0] = imgs_resized[i].shape[1]
        shapes_before_pad[i, 1] = imgs_resized[i].shape[2]
        imgs_resized[i], pad = pad_to_square(imgs_resized[i], 0)
        paddings[i, 0] = torch.tensor(pad[0]) # x left padding
        paddings[i, 1] = torch.tensor(pad[2]) # y top padding
    
    if targets is None:
        return
    ## Convert targets
    #############################################
    
    # iterate over all targets get params from corresponding img, 
    # adapt x/y labels to the padded image, do conversion
    for i in range(targets.shape[0]):
        img_idx = targets[i, 0].long().item() # first entry is image index 
        # Extract coordinates; scale up and adjust for padding
        xy_points = targets[i, 2:].reshape(3, 2) # ... adapt ... 
        # TODO: in single step
        # scale relative label coords. to coords. of (proportionally) scaled input image
        xy_points[:, 0] = xy_points[:, 0]*shapes_before_pad[img_idx, 1] + paddings[img_idx, 0]
        xy_points[:, 1] = xy_points[:, 1]*shapes_before_pad[img_idx, 0] + paddings[img_idx, 1]

      ... custom code ...
        # convert head back to relative coord.
        targets[i, :] = ...
      ...
return targets, imgs_resized
```

Cheers!
Felix

**Comment by Flova** [comment]
This has been reworked recently.
